### ISSUE TITLE
`jax.grad(jnp.abs)(0.0)` returns `1.0` — inconsistent with `jnp.sign(0.0) = 0.0`

### ISSUE THREAD

**Issue by JamesDLance** [jax-ml/jax#37820]
### Description

The gradient of `jnp.abs` at `x = 0` is `1.0`, but the mathematical definition is `d|x|/dx = sign(x)`, and `jnp.sign(0.0)` returns `0.0`. The JVP rule for `lax.abs` is therefore internally inconsistent: it does not use `lax.sign` but instead uses a `x >= 0` right-derivative convention that produces `1.0` at exactly zero.

This inconsistency can silently corrupt gradients in any model that passes zero through an absolute-value operation (e.g. L1 regularisation, `jnp.abs`-based norms, or any function that uses `abs` in its formulation).

## Reproducer

```python
import numpy as np
import jax
import jax.numpy as jnp
from jax import lax

# Standalone sign agrees with NumPy: sign(0) = 0
print("jnp.sign(0.0f32) =", float(jnp.sign(jnp.float32(0.0))))   # 0.0
print("lax.sign(0.0f32) =", float(lax.sign(jnp.float32(0.0))))    # 0.0
print("np.sign(0.0f32)  =", float(np.sign(np.float32(0.0))))       # 0.0

# But the gradient of abs at 0 gives 1.0, not sign(0) = 0.0
grad_abs = jax.grad(jnp.abs)
for x in [0.0, -0.0, 1.0, -1.0]:
    print(f"grad(abs)({x:+g}f32) = {float(grad_abs(jnp.float32(x)))}")

# The JVP directly confirms the issue
primals, tangents = jax.jvp(lax.abs, (jnp.float32(0.0),), (jnp.float32(1.0),))
print(f"jvp(lax.abs, (0.0,), (1.0,)) -> tangent = {float(tangents)}")  # 1.0
```

## Expected output

```
jnp.sign(0.0f32) = 0.0
lax.sign(0.0f32) = 0.0
np.sign(0.0f32)  = 0.0
grad(abs)(+0.0f32) = 0.0    ← should be sign(0) = 0
grad(abs)(-0.0f32) = 0.0    ← should be sign(-0) = 0 (or -0, same magnitude)
grad(abs)(+1.0f32) = 1.0
grad(abs)(-1.0f32) = -1.0
jvp(lax.abs, (0.0,), (1.0,)) -> tangent = 0.0
```

## Actual output

```
jnp.sign(0.0f32) = 0.0
lax.sign(0.0f32) = 0.0
np.sign(0.0f32)  = 0.0
grad(abs)(+0.0f32) = 1.0    ← WRONG: inconsistent with sign(0)=0
grad(abs)(-0.0f32) = 1.0    ← WRONG
grad(abs)(+1.0f32) = 1.0
grad(abs)(-1.0f32) = -1.0
jvp(lax.abs, (0.0,), (1.0,)) -> tangent = 1.0   ← WRONG
```

## Root cause

The JVP rule for `lax.abs` uses `where(x >= 0, g, -g)`, which evaluates to `g` (i.e., `1.0`) when `x = 0`. This is the right-hand derivative of `|x|` at zero. The consistent implementation would multiply by `lax.sign(x)`, which gives `0.0 * g = 0.0` at `x = 0`, matching `jnp.sign(0.0) = 0.0`.

## Impact

- L1 regularisation losses computed via `jnp.sum(jnp.abs(params))` will have incorrect subgradient at exactly-zero parameters.
- Any gradient-based optimiser that relies on the gradient of `abs` being zero at zero (to satisfy the optimality condition) will receive a non-zero update signal.
- The inconsistency is confusing: `jnp.sign(0.0)` explicitly documents the convention as zero, but `jax.grad(jnp.abs)(0.0)` silently contradicts it.

### System info (python version, jaxlib version, accelerator, etc.)

```
jax:     0.10.1.dev20260513+64f6e229b
jaxlib:  0.10.0
numpy:   2.4.4
python:  3.12.3
OS:      Linux-6.14.0-37-generic-x86_64-with-glibc2.39
backend: cpu
```

**Comment by jakevdp** [comment]
Thanks for the report – this issue is similar to #37821. The gradient of `abs` at zero is undefined, but returning `NaN` for the gradient leads to problems in practice. So we have a choice: we can return `-1.0` (the left-gradient), `+1.0` (the right-gradient), or we could perhaps choose something inbetween. Unlike the case for `maximum` and `minimum` (discussed in #37821) the midpoint gradient of `0.0` would cause problems in practice (e.g. gradient optimization could get stuck at this zero-gradient crossover value). So by convention we explicitly choose `+1.0`: https://github.com/jax-ml/jax/blob/fb17ce618c4cd5d9df0d4b59b823a9b986a01422/jax/_src/lax/lax.py#L4680-L4685

This is simply a convention, and other ways of spelling `def abs` in JAX may return different values – that would be true regardless of what convention we chose here.

In any case, this is working as intended, so I'm going to close this issue. Thanks for starting the discussion!
